kin-sdk-core/src/main/java/kin/sdk/core/Request.java

```diff
@@ -0,0 +1,79 @@
+package kin.sdk.core;
+
+
+import android.os.Handler;
+import android.os.Looper;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+public class Request<T> {
+
+    private static final ExecutorService executorService = Executors.newCachedThreadPool();
+    private final Handler mainHandler;
+    private final Callable<T> callable;
+    private boolean cancelled;
+    private boolean executed;
+    private Future<?> future;
+    private ResultCallback<T> resultCallback;
+
+    Request(Callable<T> callable) {
+        checkNotNull(callable, "callable");
+        this.callable = callable;
+        this.mainHandler = new Handler(Looper.getMainLooper());
+    }
+
+    synchronized public void run(ResultCallback<T> callback) {
+        checkBeforeRun(callback);
+        executed = true;
+        submitFuture(callable, callback);
+    }
+
+    private void checkBeforeRun(ResultCallback<T> callback) {
+        checkNotNull(callback, "callback");
+        if (executed) {
+            throw new IllegalStateException("Request already running.");
+        }
+        if (cancelled) {
+            throw new IllegalStateException("Request already cancelled.");
+        }
+    }
+
+    private void checkNotNull(Object param, String name) {
+        if (param == null) {
+            throw new IllegalArgumentException(name + " cannot be null.");
+        }
+    }
+
+    private void submitFuture(final Callable<T> callable, ResultCallback<T> callback) {
+        this.resultCallback = callback;
+        future = executorService.submit(() -> {
+            try {
+                final T result = callable.call();
+                executeOnMainThreadIfNotInterrupted(() -> resultCallback.onResult(result));
+            } catch (final Exception e) {
+                executeOnMainThreadIfNotInterrupted(() -> resultCallback.onError(e));
+            }
+        });
+    }
+
+    private synchronized void executeOnMainThreadIfNotInterrupted(Runnable runnable) {
```

Comment:
maybe rename to executeOnMainThread, should not need to imply on implementation, also there is no other option to execute the Runnable.

Reply:
renamed to 'executeOnMainThreadIfNotCancelled', following clean code principles, its more than OK to call method like that (better than comments), in this case where it's private. (I want to emphasize the implementation, in contrast to public API)